fathom/src/lang/core/typing.rs

```diff
@@ -331,25 +355,22 @@ impl<'me> Context<'me> {
 
                 // Check that the fields match the types from the type definition.
                 let mut missing_labels = Vec::new();
-                // Local environment for evaluating the field types with the
-                // values defined in the terms.
-                let mut type_locals = Locals::new();
-                for field_declaration in field_declarations.iter() {
-                    match pending_field_definitions.remove(&field_declaration.label.data) {
-                        Some(field_definition) => {
-                            // NOTE: It should be safe to evaluate the field type
-                            // because we trust that struct items have been checked.
-                            let r#type =
-                                self.eval_with_locals(&mut type_locals, &field_declaration.type_);
-                            // TODO: stop on errors
-                            self.check_type(&field_definition.term, &r#type);
-                            let value = self.eval(&field_definition.term);
 
-                            type_locals.push(value);
+                field_declarations.for_each_field(
+                    self.globals,
+                    &self.item_definitions.clone(), // FIXME: avoid clone
```

Comment:
It's kind of frustrating to have to do these clones of the item definitions - but it's required seeing as we take `self` by mutable reference in the closure, even though we don't mutate `items` at all. Perhaps we could use [`elsa::FrozenMap`](https://docs.rs/elsa/1.4.0/elsa/map/struct.FrozenMap.html) to make this a bit nicer?